### ISSUE TITLE
[Bug]: Errors with saving/removing data from secure storage

### ISSUE THREAD

**Issue by appimpact** [shinyorg/shiny#1120]
### Component/Nuget

Push - Native (Shiny.Push)

### What operating system(s) are effected?

- [X] iOS 12
- [X] iOS 13
- [X] iOS 14
- [X] iOS 15
- [ ] Mac Catalyst
- [ ] Android 8
- [ ] Android 9
- [ ] Android 10
- [ ] Android 11
- [ ] Android 12

### What platform(s) are effected?

- [ ] MAUI
- [ ] Xamarin Native
- [X] Xamarin Forms

### Steps To Reproduce

Issue is related to how is PushManager using Store Extensions:
https://github.com/shinyorg/shiny/blob/260-dev/src/Shiny.Core/Stores/StoreExtensions.cs

Issue is happening only on iOS devices.

There are three errors that we have managed to isolate:

![image](https://user-images.githubusercontent.com/34662138/196187281-fb89e1fb-4ae8-42fe-97bd-3f9e85c4bd7c.png)

Exception log for first error with "DuplicateItem":

```
SecureKeyValueStore.Set (System.String key, System.Object value)
System.ArgumentException: Failed to add secure value - DuplicateItem

Stack trace:

    Shiny.Stores
    SecureKeyValueStore.Set (System.String key, System.Object value)
    Shiny
    StoreExtensions.SetOrRemove (Shiny.Stores.IKeyValueStore store, System.String key, System.Object value)
    Shiny.Push
    PushContainer.set_CurrentRegistrationToken (System.String value)
    Shiny.Push
    PushContainer.SetCurrentToken (System.String token, System.Boolean fireChangeIfApplicable)
    Shiny.Push.FirebaseMessaging
    PushManager.RequestAccess (System.Threading.CancellationToken cancelToken)
```

Exception log for second error with "ItemNotFound":

```
SecureKeyValueStore.Remove (System.String key)
System.ArgumentException: Error removing secure value - ItemNotFound

Stack trace:

    Shiny.Stores
    SecureKeyValueStore.Remove (System.String key)
    Shiny.Stores
    SecureKeyValueStore.Set (System.String key, System.Object value)
    Shiny
    StoreExtensions.SetOrRemove (Shiny.Stores.IKeyValueStore store, System.String key, System.Object value)
    Shiny.Push
    PushContainer.set_CurrentRegistrationToken (System.String value)
    Shiny.Push
    PushContainer.SetCurrentToken (System.String token, System.Boolean fireChangeIfApplicable)
    Shiny.Push.FirebaseMessaging
    PushManager.<TryStartFirebase>b__15_1 (System.String token)
    System.Runtime.CompilerServices
    AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state)
    Foundation
    NSAsyncSynchronizationContextDispatcher.Apply ()
    (wrapper managed-to-native) UIKit.UIApplication.xamarin_UIApplicationMain(int,string[],intptr,intptr,intptr&)
    UIKit
    UIApplication.UIApplicationMain (System.Int32 argc, System.String[] argv, System.IntPtr principalClassName, System.IntPtr delegateClassName)

```

Third error with "Object reference not set to an instance of an object":

```

PushContainer.SetCurrentToken (System.String token, System.Boolean fireChangeIfApplicable)
System.NullReferenceException: Object reference not set to an instance of an object.

Stack trace:
Shiny.Push
PushContainer.SetCurrentToken (System.String token, System.Boolean fireChangeIfApplicable)
Shiny.Push.FirebaseMessaging
PushManager.<TryStartFirebase>b__15_1 (System.String token)
System.Runtime.CompilerServices
AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state)
Foundation
NSAsyncSynchronizationContextDispatcher.Apply ()
(wrapper managed-to-native) UIKit.UIApplication.xamarin_UIApplicationMain(int,string[],intptr,intptr,intptr&)
UIKit
UIApplication.UIApplicationMain (System.Int32 argc, System.String[] argv, System.IntPtr principalClassName, System.IntPtr delegateClassName)
UIKit
UIApplication.Main (System.String[] args, System.Type principalClass, System.Type delegateClass)
```

Suggestion would be to add checks to this method:
```
        public static void SetOrRemove(this IKeyValueStore store, string key, object? value)
        {
            if (value.IsNullOrDefault())
                store.Remove(key);
            else
                store.Set(key, value!);
        }

```

If key does not exist already - do not remove it.
If key already exists, update it instead of adding it again.
Third issue is most likely caused by previous two.

We cannot create reproducible sample as we are not using any custom code that interferes with how Shiny is managing storing of FCM tokens.

This is captured by App Center in production.

Any suggestions on how this can be mitigated would be appreciated.

### Expected Behavior

Errors should not be happening when key is missing or when key already exists.

### Actual Behavior

Errors are continuously happening on all devices where library is used.

### Code of Conduct

- [X] My reproducible sample is NOT FROM THE SHINY SAMPLES REPOSITORY!  If you file an issue that says 'use shiny sample', YOU WILL BE BANNED FROM THE SHINY REPOS.  Read the code of conduct for why we care about this!
- [X] I am using the latest stable version from nuget (sponsors can discuss previews in discord)
- [X] I agree to follow this project's Code of Conduct

**Comment by aritchie** [comment]
I haven't been able to reproduce these issues directly, but I have seen them in the past.  

Log 1 - I have never seen this and I have tests to cover it.
So log 2 & 3 appear to be happening on app start.  Is there a push notification being sent to the device before it unlocks for the first time?

**Comment by appimpact** [comment]
@aritchie  No push notifications have been sent to these users.

It happens as soon as they install app for the first time and opt in to receive push notifications.

Apps crash instantly as we have no way to catch and handle this exception except to log it.

After 3-4 attempts, users manage to finally start the app.

Is there a way to handle this somehow externally?

**Comment by aritchie** [comment]
I don't know what the problem is yet so I can't really provide a workaround.  I need to figure out a way to reproduce it.

**Comment by aritchie** [comment]
1) Your first log is catchable with RequestAccess
2) You can also try removing the use of secure storage by using Shiny.Stores.KeyValueStoreFactory.DefaultStoreName = "settings";

Do you have any other values in your entitlements.plist besides push setup?

**Comment by appimpact** [comment]
@aritchie  Thanks.

1. We actually are catching that one, sorry.
2. That value is already set to "settings" be default:
![image](https://user-images.githubusercontent.com/34662138/196259873-51096cbf-236c-41ca-a50d-54f8fb8fff3f.png)

Did you mean to change it to something else or?

**Comment by aritchie** [comment]
No.  You're looking at the netstandard version.  The iOS version defaults to secure.

**Comment by appimpact** [comment]
@aritchie Entitlements.plist:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>aps-environment</key>
	<string>production</string>
</dict>
</plist>

```

**Comment by appimpact** [comment]
@aritchie  Screenshot from appcenter.ms for "ItemNotFound" exception:

![image](https://user-images.githubusercontent.com/34662138/197068930-1f9a1a2b-1ffd-4041-834c-b58ad3c300c3.png)

**Comment by appimpact** [comment]
@aritchie  Screenshot for "DuplicateItem" exception:

![image](https://user-images.githubusercontent.com/34662138/197069097-11b75b43-ba2c-46e1-9b88-597964f6b8ce.png)

**Comment by aritchie** [comment]
I have tried everything to reproduce this without any luck.  I've got tests passing with the exact use cases this runs up against.  If no one can provide additional info or reproducible sample, this will be closed.

Please try the workaround with using platform preferences.  If you have a specific device having this issue, run the tests listed above and add findings
